### ISSUE TITLE
Error: Provider configuration not present when adding a new provider

### ISSUE THREAD

**Issue by sharebear** [terraform-providers/terraform-provider-azurerm#11121]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v0.14.2
+ provider registry.terraform.io/hashicorp/azurerm v2.48.0
+ provider registry.terraform.io/hashicorp/random v2.3.0
```

### Affected Resource(s)

* `azurerm_key_vault_secret`

### Terraform Configuration Files

environments/prod/main.tf
```hcl
provider "azurerm" {
  features {}
  subscription_id            = var.subscription_id
  skip_provider_registration = true
}

provider "azurerm" {
  alias = "admin"
  subscription_id            = var.admin_subscription_id
  skip_provider_registration = true
}

module "my_module" {
  source              = "../../modules/my-module"
  providers = {
    azurerm = azurerm
    azurerm.admin = azurerm.admin
  }
  # Skipped lots more for brevity
}
```

modules/my-module/main.tf

```hcl
resource "random_password" "db_administrator_password" {
  length = 128
  min_upper   = 1
  min_lower   = 1
  min_numeric = 1
  min_special = 1
}

resource "azurerm_sql_server" "my_server" {
  # removed irrelevant config
  administrator_login          = "myadmin"
  administrator_login_password = random_password.db_administrator_password.result
}

resource "azurerm_key_vault_secret" "db_administrator_password" {
  provider     = azurerm.admin
  name         = "${var.name_prefix}-db-adm-pword"
  value        = random_password.db_administrator_password.result
  key_vault_id = var.admin_vault_id

  tags = var.tags
}
```

### Debug Output

N/A can provide at a later date if important

### Panic Output

N/A

### Expected Behaviour

That the database password should be inserted in a vault available to the secondary provider.

### Actual Behaviour

```
terraform plan

Error: Provider configuration not present

To work with
module.my_module.azurerm_key_vault_secret.db_administrator_password its
original provider configuration at
module.my_module.provider["registry.terraform.io/hashicorp/azurerm"].admin
is required, but it has been removed. This occurs when a provider
configuration is removed while objects created by that provider still exist in
the state. Re-add the provider configuration to destroy
module.my_module.azurerm_key_vault_secret.db_administrator_password,
after which you can remove the provider configuration again.
```

### Steps to Reproduce

1. `terraform plan`

### Important Factoids

I'm trying to save the admin password for an Azure SQL database to a vault for easy access for administrators when necessary. During development I had the vault and SQL server in the same subscription that worked fine, but for production I have my database and the vault in separate subscriptions. This failed as while I provide the full resource id to the vault to the `azurerm_key_vault_secret` the subscription id in the resource id is ignored when calling the azure sdk (as it is bound to the subscription in the provider declaration) so it fails to find the vault at all.

I'm trying to work around this by using a second provider configuration, but I'm getting the above error message that makes little sense to me. Looking at the statefile the resource doesn't exist there so I don't understand how I'm supposed to remove it. I also don't understand how I can re-add a provider that I've only just added.

Not sure if relevant, but I'm using cli authentication when testing locally.

**Comment by favoretti** [comment]
Thank you for reporting this. Out of curiosity - did you run `terraform init` before planning with 2 providers?

**Comment by sharebear** [comment]
To be honest, I'm not certain. If I didn't do it before the first failure I think I tried running init as a possible solution. I'll try and spend some time later this week to see if I can create a minimal reproduction.

**Comment by garyburtonjisc** [comment]
@sharebear in your submodule I think you need a proxy provider config block that should stop the error you are getting but in my case it still doesn't work because then it seems to try and use the proxy provider over the passed in one. https://www.terraform.io/docs/language/modules/develop/providers.html#proxy-configuration-blocks

**Comment by sharebear** [comment]
Thanks a lot @garyburtonjisc adding the proxy provider did solve my problem. I feel that the error message could do with some work though, the large paragraph making the assumption that the provider was missing because it had been removed rather than something I was in the progress of adding, caused a lot of the confusion.

I don't have a constructive suggestion right now but I feel that this sentence

> The subdirectory ./tunnel must then contain proxy configuration blocks like the following

is tucked away at the end of the section, making it too easy to skim over when trying to rush through a fix for my failing config.

Also making the opposite error, specifying the proxy without defining providers when configuring the module provides a similarly unhelpful error message.

```
Error: "features": required field is not set
```

**Comment by sharebear** [comment]
In light of the solution I guess this issue is filed on the wrong repository. I assume the behaviour would be the same no matter which provider was being used.

**Comment by sharebear** [comment]
I see that this issue is addressed nicely in the terraform 0.15.0 release https://github.com/hashicorp/terraform/pull/27739 I guess we can just close this now.
